blog/content/post/2024-04-12-odroid-m1-secure-boot.md

```diff
@@ -0,0 +1,598 @@
+
+Now that we have pre-loader (TPL + SPL) signature verification we can use
+pre-loader to verify next steps.
+
+To complete this stage I decided to use v2024.01
+[mainline U-Boot](https://github.com/3mdeb/u-boot/tree/2024.01-odroid-m1-sb-rk3568)
+with couple changes. I also needed rkbin repository and RSA keys and certificate
+(copied into u-boot directory). It's possible to use the same key as
+earlier or create new one.
+
+### U-Boot configuration
+
+The main changes needed in my commit were just adding `signature` and
+`u-boot-spl-pubkey-dtb` node in
+[`rockchip-u-boot.dtsi`](https://github.com/3mdeb/u-boot/blob/b5f0de18708112ce61a56526e6081796045e1763/arch/arm/dts/rockchip-u-boot.dtsi)
+and changing `CONFIG_SPL_STACK_R_MALLOC_SIMPLE_LEN` config variable to
+`0x150000` to fix `alloc space exhausted` error when booting.
+
+```text
+U-Boot SPL 2024.01-dirty (Jan 01 1970 - 00:00:00 +0000)
+Trying to boot from MMC2
+alloc space exhausted
+FIT buffer size: 1199104 bytes
+Could not get FIT buffer of 1199104 bytes
+        check CONFIG_SPL_SYS_MALLOC_SIZE
+```
+
+To configure U-Boot it's enough to use `odroid-m1-sb-rk3568_defconfig` and set
+couple variables:
+
+```shell
+export CROSS_COMPILE=aarch64-linux-gnu-
+export BL31=<path/to/rkbin>/bin/rk35/rk3568_bl31_v1.44.elf
+export ROCKCHIP_TPL=<path/to/rkbin>/bin/rk35/rk3568_ddr_1560MHz_v1.21.bin
+```
+
+This time I used `gcc-aarch64-linux-gnu` cross-compiler from Debian package
+manager.
+
+### Build
+
+After configuration, we build by using `make`. It should build signed U-Boot with
+public key embedded inside SPL.
+
+```text
+make -j$(nproc)
+(...)
+ Default Configuration: 'config-1'
+ Configuration 0 (config-1)
+  Description:  rk3568-odroid-m1.dtb
+  Kernel:       unavailable
+  Firmware:     atf-1
+  FDT:          fdt-1
+  Loadables:    u-boot
+                atf-2
+                atf-3
+                atf-4
+                atf-5
+                atf-6
+  Sign algo:    sha256,rsa2048:dev
+  Sign value:   c05e2589eb863bfb9b5b5c87ea9f99042c4cb78797a6cd0d772a68bfac21bc621790cb977bd414928b7b88a50273d336aea4afb7a1a4008834e1c3c18eb08fb617bb71205e6773904f42e81364a15eb5c581425f22d6be5e30dd1a44cbe7fa626b7fb34a3eb742c03b0ec47c19dd957358949764f82ffb24f9359efe083ea04ada512f8cab0f469cac28d08da3dfffbde1516cb225b3011f36495c959793d11b8c4bde234caca5ac1d779f7983dcf865be92c2ac3300e4cdf536e51fa398c8930dafbdd0a3258ba4704eebc063e5e57533a962a2da5c7eaf447b1244f9720adb3dc775d95b4c6dc99f6b9bd73ebe992095429510b86dd5ac912c24a8cec64841
+  Timestamp:    Mon May 27 15:08:11 2024
+Signature written to 'u-boot.itb', node '/configurations/config-1/signature'
+  OFCHK   .config
+```
+
+### Signing idbloader
+
+Signing idbloader is similar to [Signing Loader](#signing-loader) section except
+with `sb --idb` argument.
+It's important to remember to sign idbloader with the same keys as used in that
+section (in case current ones are different).
+
+```shell
+tools/rk_sign_tool sb --idb ../u-boot/idbloader.img
```

Comment:
Could this be verified without booting?

Reply:
f7982c42f7d027003915f05f61168e15a1a64921